### PR TITLE
Introduce validation_options

### PR DESCRIPTION
**Pull request by fermaem**
Following https://github.com/SkyLothar/lua-resty-jwt/pull/18#issuecomment-207060867 a first spike at it.

### CODE REVIEW

**Comment by fermaem** [comment]
@SkyLothar Although this PR is (obviously) not ready yet, I'd be very grateful if you could provide me with a first feedback (at least from the API perspective).

**Comment by fermaem** [comment]
Still missing a ton of unit tests, but at least the build is passing

**Comment by SkyLothar** [comment]
:cool: :exclamation: 

`validation options table` is simple and elegant. but how about a `validator table`.
```lua
function my_iss_validator(whitelist)
    function iss_validator(jwt_obj)
        local iss = jwt_obj['header']['iss']
        return has_value(whitelist, iss)
    end
end
jwt:add_validators({'nbf'=standard_nbf_validator(leeway=30), 'iss'=my_iss_validator({'iss-0', 'iss-1'})})
```

1. better code structure, no more giant all-in-one validate function. To support a new claim, just add a validator to the default validator map.
2. user can perform complex validation, iss-0's `validity_grace_period` is 100, and iss-1's is 60.

reason 2 sounds like it's over designed, what's your opinion?

**Comment by fermaem** [comment]
Thanks a lot for this feedback!

>how about a validator table.

I thought about something along those lines. However, I don't think that this should apply to all kind of validations and I feel very wary about letting the user build this validator table. My feeling is two fold:

 - Properly validating a claim requires to actually read the spec with some security oriented mindset. I'm not sure that every user of the library would be ready (or even able to do it securely). So allowing the replacement of the default claim validators seems pushing too much responsibility on the user's shoulders.
 - I strongly believe in simple, safe by default systems. Systems that can be used without spending too much time reading the configuration while offering a reasonable sets of already configured behaviors (I named the function `normalize_validation_options` with that in mind. For instance, I'd like to propose that, grace period is set to 0 second by default, exp and nbf are checked by default when present. Of course, those defaults could be of course turned down by the user, through an explicit change in the options, or made even stricter with `requires_nbf`, `requires_exp` additional options...)

However, some claims may require (or optionally expose a way to retrieve) some additional input from the user. We could provide this through a callback mechanism.

>better code structure, no more giant all-in-one validate function. 

I'm also bothered by this long function and agree with you that it should be refactored. However, I'm still not there yet. I'd like first to add some unit test coverage and maybe some additional claim validators. Then, maybe a pattern will emerge that will drive the refactoring.

Maybe this validator table would eventually emerge as an internal orchestration system, created by the processing of `normalize_validation_options`..

Thoughts?

**Comment by SkyLothar** [comment]
Of cause this lib should be `easy to use`, `battery included`
By default, we provide a `validator table` that check all standard claims when presented.
But we also provide a set of functions that can be easily used to create your own validator.

So, in my mind, the default validator table should be
```lua
{
  exp=standard_exp_validator(),
  nbf=standard_nbf_validator(),
  iss=...,
  ...
}
```
user can still use it without any configuration.

if user want change `exp`'s `validity_grace_period` to 10, and do something after the `exp` validation.
```lua
function do_something()
end
jwt:override_validator({exp=standard_exp_validator(10, do_something)})
```

----
we agreed on
  1. `conversion over configuration`, user can leave the validator table untouched. lib will provided a reasonable claim validation. 
  2. `easy to use`, by reading a table of `validation options`, user can quickly tweak the validation.( `normalize_validation_options` will create the `validator table`)

----
if we give experienced user the power to override the `validator table`
  1. by changing the input of validator creator function, user can tweak or add callback on any claim they want,
  2. add their own complex validator


:cool: let's add add more tests!

File: lib/resty/jwt.lua [hunk 1]
```diff
@@ -506,11 +537,44 @@ local function validate_exp_nbf(jwt_obj, is_validation_loose, leeway)
...
+--@function validate issuers - ensure issuer belong to a whitelist
+--@param jwt_obj, validation_options
+local function validate_iss(jwt_obj, validation_options)
+  if jwt_obj[str_const.reason] ~= nil then
+    return
+  end
+
+  local issuer = jwt_obj[str_const.payload][str_const.iss]
+
+  if iss == nil then
```
Comment (SkyLothar) [inline lib/resty/jwt.lua]: you mean `issuer` i think.

File: lib/resty/jwt.lua [hunk 2]
```diff
@@ -506,11 +537,44 @@ local function validate_exp_nbf(jwt_obj, is_validation_loose, leeway)
...
+  end
+
+  local valid_issuers = validation_options[str_const.valid_issuers]
+
+  if valid_issuers == nil then
+    return
+  end
+
+  for k in pairs(valid_issuers) do
+    if iss == k then
```
Comment (SkyLothar) [inline lib/resty/jwt.lua]: or you can change L574 to `iss`, `k` is a bad name for variable.

**Comment by SkyLothar** [comment]
We should write tests before add new functions :beetle:

**Comment by fermaem** [comment]
> We should write tests before add new functions

@SkyLothar Done and ready for a first review.

For the sake of the exercise I've converted the lifetime in a validator and added the `iss` content validator.

I've also moved the existing tests that were targeting claim validation in separate files.

File: lib/resty/jwt.lua [hunk 3]
```diff
@@ -464,43 +502,84 @@ function _M.load_jwt(self, jwt_str, secret)
...
-    return
-  end
-
+--@function validate exp nbf claims - validate expiry and not valid before
+--@param jwt_obj, leeway, require_nbf_claim, require_exp_claim
+local function validate_lifetime(jwt_obj, leeway, require_nbf_claim, require_exp_claim)
   local exp = jwt_obj[str_const.payload][str_const.exp]
   local nbf = jwt_obj[str_const.payload][str_const.nbf]
 
+  if (leeway ~= nil and exp == nil and nbf == nil) then
```
Comment (SkyLothar) [inline lib/resty/jwt.lua]: by default, leeway(`lifetime_grace_period` is 0),
so, a jwt with without lifetime claims is an invalid jwt?

maybe we should set default `lifetime_grace_period` to `nil`?

Comment (nulltoken) [inline lib/resty/jwt.lua]: > by default, leeway(lifetime_grace_period is 0),
> so, a jwt with without lifetime claims is an invalid jwt?

Indeed, if both of the `exp` and `nbf` claims are missing, the lifetime cannot be validated.

> maybe we should set default lifetime_grace_period to nil?

You're right. I'm going to fix this.

Comment (fermaem) [inline lib/resty/jwt.lua]: > by default, leeway(lifetime_grace_period is 0),
> so, a jwt with without lifetime claims is an invalid jwt?

Indeed, if both of the `exp` and `nbf` claims are missing, the lifetime cannot be validated.

> maybe we should set default lifetime_grace_period to nil?

You're right. I'm going to fix this.

Comment (fermaem) [inline lib/resty/jwt.lua]: Fixed

**Comment by fermaem** [comment]
Also updated the `README.md` with some documentation

**Comment by SkyLothar** [comment]
Cool
Thanks

**Pull request merged**